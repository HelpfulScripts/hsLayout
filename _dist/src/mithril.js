if (!global['window']) {
    console.log('creating non-browser polyfill');
    global['window'] = require("mithril/test-utils/browserMock.js")();
    global['document'] = window.document;
}
export const m = require("mithril");
const o = require("mithril/ospec/ospec");
o.root = window.document.createElement("div");
export { o };
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoibWl0aHJpbC5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbIi4uLy4uL3NyYy9taXRocmlsLnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUlBLElBQUksQ0FBQyxNQUFNLENBQUMsUUFBUSxDQUFDLEVBQUU7SUFDbkIsT0FBTyxDQUFDLEdBQUcsQ0FBQywrQkFBK0IsQ0FBQyxDQUFDO0lBRTdDLE1BQU0sQ0FBQyxRQUFRLENBQUMsR0FBRyxPQUFPLENBQUMsbUNBQW1DLENBQUMsRUFBRSxDQUFDO0lBQ2xFLE1BQU0sQ0FBQyxVQUFVLENBQUMsR0FBRyxNQUFNLENBQUMsUUFBUSxDQUFDO0NBRXhDO0FBS0QsTUFBTSxDQUFDLE1BQU0sQ0FBQyxHQUFHLE9BQU8sQ0FBQyxTQUFTLENBQUMsQ0FBQztBQU9wQyxNQUFNLENBQUMsR0FBRyxPQUFPLENBQUMscUJBQXFCLENBQUMsQ0FBQztBQUN6QyxDQUFDLENBQUMsSUFBSSxHQUFHLE1BQU0sQ0FBQyxRQUFRLENBQUMsYUFBYSxDQUFDLEtBQUssQ0FBQyxDQUFDO0FBRTlDLE9BQU8sRUFBRSxDQUFDLEVBQUUsQ0FBQyJ9