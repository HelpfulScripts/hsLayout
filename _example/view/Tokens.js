export class LayoutToken {
    constructor(size) {
        this.size = size;
    }
    getSize() { return this.size; }
}
export class DefinedToken extends LayoutToken {
    constructor(size) { super(size); }
}
export class FillToken extends LayoutToken {
    constructor() { super(-1); }
}
export class PixelToken extends DefinedToken {
    constructor(size) { super(size); }
}
export class PercentToken extends DefinedToken {
    constructor(size) { super(size); }
}
export function px(px) { return new PixelToken(px); }
export function pc(pc) { return new PercentToken(pc); }
export const FILL = new FillToken();
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiVG9rZW5zLmpzIiwic291cmNlUm9vdCI6IiIsInNvdXJjZXMiOlsiLi4vLi4vc3JjL3ZpZXcvVG9rZW5zLnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQVFBLE1BQU07SUFDRixZQUFvQixJQUFZO1FBQVosU0FBSSxHQUFKLElBQUksQ0FBUTtJQUFHLENBQUM7SUFDN0IsT0FBTyxLQUFLLE9BQU8sSUFBSSxDQUFDLElBQUksQ0FBQyxDQUFDLENBQUM7Q0FDekM7QUFLRCxNQUFNLG1CQUE2QixTQUFRLFdBQVc7SUFDbEQsWUFBWSxJQUFZLElBQUksS0FBSyxDQUFDLElBQUksQ0FBQyxDQUFDLENBQUMsQ0FBQztDQUM3QztBQUtELE1BQU0sZ0JBQWlCLFNBQVEsV0FBVztJQUN0QyxnQkFBZ0IsS0FBSyxDQUFDLENBQUMsQ0FBQyxDQUFDLENBQUMsQ0FBQyxDQUFDO0NBQy9CO0FBS0QsTUFBTSxpQkFBa0IsU0FBUSxZQUFZO0lBQ3hDLFlBQVksSUFBVyxJQUFJLEtBQUssQ0FBQyxJQUFJLENBQUMsQ0FBQyxDQUFDLENBQUM7Q0FDNUM7QUFLRCxNQUFNLG1CQUFvQixTQUFRLFlBQVk7SUFDMUMsWUFBWSxJQUFXLElBQUksS0FBSyxDQUFDLElBQUksQ0FBQyxDQUFDLENBQUMsQ0FBQztDQUM1QztBQU1ELE1BQU0sYUFBYSxFQUFTLElBQU0sT0FBTyxJQUFJLFVBQVUsQ0FBQyxFQUFFLENBQUMsQ0FBQyxDQUFDLENBQUM7QUFNOUQsTUFBTSxhQUFhLEVBQVMsSUFBTSxPQUFPLElBQUksWUFBWSxDQUFDLEVBQUUsQ0FBQyxDQUFDLENBQUMsQ0FBQztBQUtoRSxNQUFNLENBQUMsTUFBTSxJQUFJLEdBQUcsSUFBSSxTQUFTLEVBQUUsQ0FBQyJ9