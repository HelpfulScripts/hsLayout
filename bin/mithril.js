"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.m = require("mithril");
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoibWl0aHJpbC5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbIi4uL3NyYy9taXRocmlsLnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7O0FBUWEsUUFBQSxDQUFDLEdBQUcsT0FBTyxDQUFDLFNBQVMsQ0FBQyxDQUFDIn0=