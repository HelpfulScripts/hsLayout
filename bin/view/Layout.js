"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mithril_1 = require("../mithril");
const Layouter_1 = require("./Layouter");
class Layout {
    getComponents(node) {
        return !Array.isArray(node.attrs.content) ? node.attrs.content :
            node.attrs.content.map((c) => {
                if (c.compClass) {
                    c.attrs.route = node.attrs.route;
                    return mithril_1.m(c.compClass, c.attrs);
                }
                else {
                    return c;
                }
            });
    }
    getCSS(node) {
        return node.attrs.css || '';
    }
    normalizeContent(components) {
        if (typeof components === 'string') {
            return [mithril_1.m('.hs-leaf', mithril_1.m.trust(components))];
        }
        if (components.length > 0) {
            return components.map((comp) => (comp instanceof Layout) ? comp : mithril_1.m(Layout, { content: comp }));
        }
        return [components];
    }
    view(node) {
        const content = this.normalizeContent(this.getComponents(node));
        let css = Layouter_1.Layouter.createLayout(node.attrs, content);
        const attrs = {
            style: node.style,
            route: node.attrs.route,
            onclick: node.attrs.onclick
        };
        node.attrs.route = undefined;
        if (node.attrs.href) {
            attrs.href = node.attrs.href;
            attrs.oncreate = mithril_1.m.route.link;
            attrs.onupdate = mithril_1.m.route.link;
        }
        return mithril_1.m(`.hs-layout ${css} ${this.getCSS(node)}`, attrs, content.map((c) => c));
    }
}
exports.Layout = Layout;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiTGF5b3V0LmpzIiwic291cmNlUm9vdCI6IiIsInNvdXJjZXMiOlsiLi4vLi4vc3JjL3ZpZXcvTGF5b3V0LnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7O0FBaUJBLHdDQUEwQztBQUMxQyx5Q0FBMEM7QUF5QjFDO0lBb0JjLGFBQWEsQ0FBQyxJQUFVO1FBQzlCLE9BQU8sQ0FBQyxLQUFLLENBQUMsT0FBTyxDQUFDLElBQUksQ0FBQyxLQUFLLENBQUMsT0FBTyxDQUFDLENBQUEsQ0FBQyxDQUFDLElBQUksQ0FBQyxLQUFLLENBQUMsT0FBTyxDQUFDLENBQUM7WUFDM0QsSUFBSSxDQUFDLEtBQUssQ0FBQyxPQUFPLENBQUMsR0FBRyxDQUFDLENBQUMsQ0FBSyxFQUFFLEVBQUU7Z0JBQzdCLElBQUksQ0FBQyxDQUFDLFNBQVMsRUFBRTtvQkFDYixDQUFDLENBQUMsS0FBSyxDQUFDLEtBQUssR0FBRyxJQUFJLENBQUMsS0FBSyxDQUFDLEtBQUssQ0FBQztvQkFDakMsT0FBTyxXQUFDLENBQUMsQ0FBQyxDQUFDLFNBQVMsRUFBRSxDQUFDLENBQUMsS0FBSyxDQUFDLENBQUM7aUJBQ2xDO3FCQUFNO29CQUNILE9BQU8sQ0FBQyxDQUFDO2lCQUNaO1lBQ0wsQ0FBQyxDQUFDLENBQUM7SUFDWCxDQUFDO0lBUVMsTUFBTSxDQUFDLElBQVU7UUFDdkIsT0FBTyxJQUFJLENBQUMsS0FBSyxDQUFDLEdBQUcsSUFBSSxFQUFFLENBQUM7SUFDaEMsQ0FBQztJQUdPLGdCQUFnQixDQUFDLFVBQTZDO1FBQ2xFLElBQUksT0FBTyxVQUFVLEtBQUssUUFBUSxFQUFFO1lBQ2hDLE9BQU8sQ0FBQyxXQUFDLENBQUMsVUFBVSxFQUFFLFdBQUMsQ0FBQyxLQUFLLENBQUMsVUFBVSxDQUFDLENBQUMsQ0FBQyxDQUFDO1NBQy9DO1FBQ0QsSUFBSSxVQUFVLENBQUMsTUFBTSxHQUFDLENBQUMsRUFBRTtZQUNyQixPQUFPLFVBQVUsQ0FBQyxHQUFHLENBQUMsQ0FBQyxJQUF5QixFQUFRLEVBQUUsQ0FDbEQsQ0FBQyxJQUFJLFlBQVksTUFBTSxDQUFDLENBQUEsQ0FBQyxDQUFDLElBQUksQ0FBQyxDQUFDLENBQUMsV0FBQyxDQUFDLE1BQU0sRUFBRSxFQUFDLE9BQU8sRUFBQyxJQUFJLEVBQUMsQ0FBQyxDQUNqRSxDQUFDO1NBQ0w7UUFFRCxPQUFPLENBQUMsVUFBVSxDQUFDLENBQUM7SUFDeEIsQ0FBQztJQXFCRCxJQUFJLENBQUMsSUFBVTtRQUNYLE1BQU0sT0FBTyxHQUFHLElBQUksQ0FBQyxnQkFBZ0IsQ0FBQyxJQUFJLENBQUMsYUFBYSxDQUFDLElBQUksQ0FBQyxDQUFDLENBQUM7UUFDaEUsSUFBSSxHQUFHLEdBQUcsbUJBQVEsQ0FBQyxZQUFZLENBQUMsSUFBSSxDQUFDLEtBQUssRUFBRSxPQUFPLENBQUMsQ0FBQztRQUNyRCxNQUFNLEtBQUssR0FBTztZQUNkLEtBQUssRUFBRSxJQUFJLENBQUMsS0FBSztZQUNqQixLQUFLLEVBQUUsSUFBSSxDQUFDLEtBQUssQ0FBQyxLQUFLO1lBQ3ZCLE9BQU8sRUFBRSxJQUFJLENBQUMsS0FBSyxDQUFDLE9BQU87U0FDOUIsQ0FBQztRQUNGLElBQUksQ0FBQyxLQUFLLENBQUMsS0FBSyxHQUFHLFNBQVMsQ0FBQztRQUM3QixJQUFJLElBQUksQ0FBQyxLQUFLLENBQUMsSUFBSSxFQUFFO1lBQ2pCLEtBQUssQ0FBQyxJQUFJLEdBQUcsSUFBSSxDQUFDLEtBQUssQ0FBQyxJQUFJLENBQUM7WUFDN0IsS0FBSyxDQUFDLFFBQVEsR0FBRyxXQUFDLENBQUMsS0FBSyxDQUFDLElBQUksQ0FBQztZQUM5QixLQUFLLENBQUMsUUFBUSxHQUFHLFdBQUMsQ0FBQyxLQUFLLENBQUMsSUFBSSxDQUFDO1NBQ2pDO1FBQ0QsT0FBTyxXQUFDLENBQUMsY0FBYyxHQUFHLElBQUksSUFBSSxDQUFDLE1BQU0sQ0FBQyxJQUFJLENBQUMsRUFBRSxFQUFFLEtBQUssRUFBRSxPQUFPLENBQUMsR0FBRyxDQUFDLENBQUMsQ0FBSyxFQUFFLEVBQUUsQ0FBQyxDQUFDLENBQUMsQ0FBQyxDQUFDO0lBQ3pGLENBQUM7Q0FDSjtBQTNGRCx3QkEyRkMifQ==