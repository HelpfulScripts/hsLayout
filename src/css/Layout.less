@LayoutSpacing: 1px solid white;

.hs-layout {
    left: 0%; top: 0%;
    width:100%; 
    height:100%;
    position: absolute; 
    display: block;
    overflow: scroll;
    border: 0;
}

.hs-layout-spacing {
    .hs-column-layout & > .hs-layout {
        border: @LayoutSpacing;
        & > .hs-layout:first-of-type { border-left: 0; }  
        & > .hs-layout:last-of-type { border-right: 0; }  
    } 
    .hs-row-layout {
        & > .hs-layout { border: @LayoutSpacing; }
        & > .hs-layout:first-of-type { border-top: 0; }  
        & > .hs-layout:last-of-type { border-bottom: 0; }  
    } 
}

.hs-row-layout {
    box-sizing: border-box;
    &>.hs-layout { box-sizing: border-box; }
}

.hs-column-layout {
    box-sizing: border-box;
    &>.hs-layout { box-sizing: border-box; }
}

.hs-tile-layout>.hs-layout {
    box-sizing: border-box;
}

.hs-layout-frame {
    position:relative;
}

.hs-layout-fill {
    position:relative;
    padding:0;
    margin:0;
    width: 100%;
    height: 100%;
}

.hs-align-left {
    text-align: left;
}

.hs-align-right {
    text-align: right;
}

.hs-align-center {
    text-align: center;
}
