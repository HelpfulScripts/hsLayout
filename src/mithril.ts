/**
 * translates mithril libraries to an ES6 module and provides some Typescript type shortcuts.
 */


/**
 * import and re-export the mithril m objkect
 */
import _m from "mithril";

export const m = _m;
export type Vnode = _m.Vnode<any, any>;
