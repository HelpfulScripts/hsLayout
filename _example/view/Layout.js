import { m } from '../mithril';
import { Layouter } from './Layouter';
export class Layout {
    getComponents(node) {
        return !Array.isArray(node.attrs.content) ? node.attrs.content :
            node.attrs.content.map((c) => {
                if (c.compClass) {
                    c.attrs.route = node.attrs.route;
                    return m(c.compClass, c.attrs);
                }
                else {
                    return c;
                }
            });
    }
    getCSS(node) {
        return node.attrs.css || '';
    }
    normalizeContent(components) {
        if (typeof components === 'string') {
            return [m('.hs-leaf', m.trust(components))];
        }
        if (components.length > 0) {
            return components.map((comp) => (comp instanceof Layout) ? comp : m(Layout, { content: comp }));
        }
        return [components];
    }
    view(node) {
        const content = this.normalizeContent(this.getComponents(node));
        let css = Layouter.createLayout(node.attrs, content);
        const attrs = {
            style: node.style,
            route: node.attrs.route,
            onclick: node.attrs.onclick
        };
        node.attrs.route = undefined;
        if (node.attrs.href) {
            attrs.href = node.attrs.href;
            attrs.oncreate = m.route.link;
            attrs.onupdate = m.route.link;
        }
        return m(`.hs-layout ${css} ${this.getCSS(node)}`, attrs, content.map((c) => c));
    }
}
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiTGF5b3V0LmpzIiwic291cmNlUm9vdCI6IiIsInNvdXJjZXMiOlsiLi4vLi4vc3JjL3ZpZXcvTGF5b3V0LnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQWlCQSxPQUFPLEVBQUUsQ0FBQyxFQUFRLE1BQVcsWUFBWSxDQUFDO0FBQzFDLE9BQU8sRUFBRSxRQUFRLEVBQUUsTUFBWSxZQUFZLENBQUM7QUF5QjVDLE1BQU07SUFvQlEsYUFBYSxDQUFDLElBQVU7UUFDOUIsT0FBTyxDQUFDLEtBQUssQ0FBQyxPQUFPLENBQUMsSUFBSSxDQUFDLEtBQUssQ0FBQyxPQUFPLENBQUMsQ0FBQSxDQUFDLENBQUMsSUFBSSxDQUFDLEtBQUssQ0FBQyxPQUFPLENBQUMsQ0FBQztZQUMzRCxJQUFJLENBQUMsS0FBSyxDQUFDLE9BQU8sQ0FBQyxHQUFHLENBQUMsQ0FBQyxDQUFLLEVBQUUsRUFBRTtnQkFDN0IsSUFBSSxDQUFDLENBQUMsU0FBUyxFQUFFO29CQUNiLENBQUMsQ0FBQyxLQUFLLENBQUMsS0FBSyxHQUFHLElBQUksQ0FBQyxLQUFLLENBQUMsS0FBSyxDQUFDO29CQUNqQyxPQUFPLENBQUMsQ0FBQyxDQUFDLENBQUMsU0FBUyxFQUFFLENBQUMsQ0FBQyxLQUFLLENBQUMsQ0FBQztpQkFDbEM7cUJBQU07b0JBQ0gsT0FBTyxDQUFDLENBQUM7aUJBQ1o7WUFDTCxDQUFDLENBQUMsQ0FBQztJQUNYLENBQUM7SUFRUyxNQUFNLENBQUMsSUFBVTtRQUN2QixPQUFPLElBQUksQ0FBQyxLQUFLLENBQUMsR0FBRyxJQUFJLEVBQUUsQ0FBQztJQUNoQyxDQUFDO0lBR08sZ0JBQWdCLENBQUMsVUFBNkM7UUFDbEUsSUFBSSxPQUFPLFVBQVUsS0FBSyxRQUFRLEVBQUU7WUFDaEMsT0FBTyxDQUFDLENBQUMsQ0FBQyxVQUFVLEVBQUUsQ0FBQyxDQUFDLEtBQUssQ0FBQyxVQUFVLENBQUMsQ0FBQyxDQUFDLENBQUM7U0FDL0M7UUFDRCxJQUFJLFVBQVUsQ0FBQyxNQUFNLEdBQUMsQ0FBQyxFQUFFO1lBQ3JCLE9BQU8sVUFBVSxDQUFDLEdBQUcsQ0FBQyxDQUFDLElBQXlCLEVBQVEsRUFBRSxDQUNsRCxDQUFDLElBQUksWUFBWSxNQUFNLENBQUMsQ0FBQSxDQUFDLENBQUMsSUFBSSxDQUFDLENBQUMsQ0FBQyxDQUFDLENBQUMsTUFBTSxFQUFFLEVBQUMsT0FBTyxFQUFDLElBQUksRUFBQyxDQUFDLENBQ2pFLENBQUM7U0FDTDtRQUVELE9BQU8sQ0FBQyxVQUFVLENBQUMsQ0FBQztJQUN4QixDQUFDO0lBcUJELElBQUksQ0FBQyxJQUFVO1FBQ1gsTUFBTSxPQUFPLEdBQUcsSUFBSSxDQUFDLGdCQUFnQixDQUFDLElBQUksQ0FBQyxhQUFhLENBQUMsSUFBSSxDQUFDLENBQUMsQ0FBQztRQUNoRSxJQUFJLEdBQUcsR0FBRyxRQUFRLENBQUMsWUFBWSxDQUFDLElBQUksQ0FBQyxLQUFLLEVBQUUsT0FBTyxDQUFDLENBQUM7UUFDckQsTUFBTSxLQUFLLEdBQU87WUFDZCxLQUFLLEVBQUUsSUFBSSxDQUFDLEtBQUs7WUFDakIsS0FBSyxFQUFFLElBQUksQ0FBQyxLQUFLLENBQUMsS0FBSztZQUN2QixPQUFPLEVBQUUsSUFBSSxDQUFDLEtBQUssQ0FBQyxPQUFPO1NBQzlCLENBQUM7UUFDRixJQUFJLENBQUMsS0FBSyxDQUFDLEtBQUssR0FBRyxTQUFTLENBQUM7UUFDN0IsSUFBSSxJQUFJLENBQUMsS0FBSyxDQUFDLElBQUksRUFBRTtZQUNqQixLQUFLLENBQUMsSUFBSSxHQUFHLElBQUksQ0FBQyxLQUFLLENBQUMsSUFBSSxDQUFDO1lBQzdCLEtBQUssQ0FBQyxRQUFRLEdBQUcsQ0FBQyxDQUFDLEtBQUssQ0FBQyxJQUFJLENBQUM7WUFDOUIsS0FBSyxDQUFDLFFBQVEsR0FBRyxDQUFDLENBQUMsS0FBSyxDQUFDLElBQUksQ0FBQztTQUNqQztRQUNELE9BQU8sQ0FBQyxDQUFDLGNBQWMsR0FBRyxJQUFJLElBQUksQ0FBQyxNQUFNLENBQUMsSUFBSSxDQUFDLEVBQUUsRUFBRSxLQUFLLEVBQUUsT0FBTyxDQUFDLEdBQUcsQ0FBQyxDQUFDLENBQUssRUFBRSxFQUFFLENBQUMsQ0FBQyxDQUFDLENBQUMsQ0FBQztJQUN6RixDQUFDO0NBQ0oifQ==