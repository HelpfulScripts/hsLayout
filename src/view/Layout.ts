/**
 * # Layout
 * A `mithril` component class that layouts available space in the window.
 * 
 * ### Invocation
 * invoked as `m(Layout, {<layout style>:[<string>], content:Array<Vnode>})`. See example below.
 * 
 * ### Attributes (node.attrs):
 * - `<layout-style>`: [...]: a required keyword indicating the {@link view.Layouter Layouter} to use, followed by a configuration array. 
 *    E.g. `rows: ["30px", "fill"]`.
 * - `content`: Array<Vnode>, required. The Vnode children to lay out. 
 *     - `string` literal, e.g. "the content"
 *     - `[ ]` array of elements:
 *         - `string` literal, e.g. "the content"
 *         - `{ }` literal
 *            - `compClass`: a {@link view.Layout.Component Component} class definition, e.g. `Layout`
 * - `css`: optional String. The css specifier to use for this `Layout` component.
 * - `route`: optional object literal holding parameters passed from `m.route`
 * - `href`: optional String. If present, makes the component clickable
 * - `target`: optional string, determines where `href` is opened. Defaults to '_blank' to open `href` in a new tab.
 * - `onclick`:(), optional. The function to call when clicked
 * 
 * ### Example
 * <example>
 * <file name='script.js'>
 * m.mount(root, {view: () => m(hsLayout.Layout, {
 *     css:     'myColumn',
 *     rows:    ["50px", "fill"], 
 *     content: ['Top row: 50px', 'Bottom row: remainder']
 *     })
 * });
 * </file>
 * </example>

 */

/** */
import { m, Vnode } from '../mithril'; 
import { Layouter } from './Layouter'; 
import { log as _log }  from 'hsutil'; const log = _log('Layout');

/**
 * Component interface. 
 * Formalizes the `mithril` requirement for a `view` method.
 */
export interface Component {
    view(node:Vnode): Vnode;
}

/**
Base class for applying layouts. Subclasses should implement a {@link view.Layout.Layout.getComponents `getComponents`} method that returns
the components to render. The default implementation returns the conponents passed in `node.attrs.content`.
Optionally, the subclass can also implement {@link view.Layout.Layout.getCSS `getCSS`} to provide the CSS class to 
assign to the component, and override the default implementation, which returns `node.attrs.css`. 
### Example:
<code>
import { Layout, px, FILL }  from 'hslayout';
const TitleHeight   = px(30); 
const FooterHeight  = px(10); 
class MyLayout extends Layout {
    getComponents(node:Vnode):Vnode {
        return this.layout('.my-layout', { rows:[TitleHeight, FILL, FooterHeight] }, [
            m(), 
            m(),
            m()
        ]);
    }
    getCSS(node:Vnode):string {
    }
} 
</code>
 */
export class Layout implements Component {
    /**
     * holds structural elements in style form: left, right, top, bottom, width, height
     */
    public style:string;

//    oninit(node:Vnode)   { this.report('Layout:init', node); }
//    oncreate(node:Vnode) { this.report('Layout:create', node); } 
//    onupdate(node:Vnode) { this.report('Layout:update', node); }

    /**
     * Called during the lifecycle `view` call to retrieve the subcomponents to render in this container.
     * The default implementation returns components stored in `node.attrs.content`. This allows for 
     * creating containers directly via mithril: `m(Layout, {content:[...]})`. 
     * In case `node.attrs.content` is an array of literals with a `compClass` field describing a Component class, 
     * the method will create a Mithril node on that class and pass the `node.attrs.route` argument down to it.
     * 
     * Override this method to create containers that return more sophisticated content.
     * @return a String, a Vnode, or an array of Strings or Vnodes
     */
    protected getComponents(node:Vnode):Vnode {
        return !Array.isArray(node.attrs.content)? node.attrs.content :
            node.attrs.content.map((c:any) => {
                if (c.compClass) { 
                    c.attrs.route = node.attrs.route;
                    return m(c.compClass, c.attrs);
                } else {
                    return c;
                }
            });
    }

    /**
     * Called during the lifecycle `view` call to retrieve the css style class to apply to this container.
     * The default implementation returns components stored in `node.attrs.css`. This allows for 
     * creating containers directly via mithril: `m(Layout, {content:[...], css:'.my-class'})`.
     * Override this method to create containers that return more sophisticated content.
     */
    protected getCSS(node:Vnode):string {
        return node.attrs.css || '';
    }

    /**
     * ensures that all `components` are `mithril` nodes and returns them in an array.
     * @param components either of:
     * - a `string` literal --> interprets the string as trusted HTML and returns a leaf node.
     * - an array of elements, each either a `Layout`, or `string` literal which will be converted to a `Layout`.
     */
    private normalizeContent(components:Array<typeof Layout|string>|string): Vnode {
        if (typeof components === 'string') { 
            return [m('.hs-leaf', m.trust(components))]; 
        }
        if (components.length>0) { // an array: ensure elements are Layout components
            return components.map((comp:string|typeof Layout):Vnode => 
                    (comp instanceof Layout)? comp : m(Layout, {content:comp})
            );
        }
        // else: assume components is a mithril node: return node as an array
        return [components];
    }

    /**
    lays out the component in `components` according to the configuration in `attrs`.
    The method returns a vnode container that has an associated `cssClass` style.
    `layout` is called during the `render` phase of the `mithril` lifecycle, 
    which ensures an outside-in calling sequence on containers; 
    i.e. the outermost containers are called first, and `node` will already have the 
    `style` field set with required style attributes. 
    These are added to any `attrs` parameter provided.

    The format for the layout configuration in `attrs` is <code>
    {<keyword>: <parameter>}
    </code>
     where `keyword` is the keyword with which the `Layouter` was registered.
     * @param cssClass a css style designator; same as used in m(cssClass, ...) 
     * @param layout the attribute object literal that configures the layout
     * @param components the components to layout within the container. 
     * This is either a primitive `string`, or an array of `Layout`s or `string`s.
     * @return a vnode that has an associated `cssClass` style.
     */
    view(node:Vnode): Vnode {
        const content = this.normalizeContent(this.getComponents(node)); // --> Vnode[]
        let css = Layouter.createLayout(node.attrs, content);
        const attrs:any = {
            style: node.style,
            route: node.attrs.route,     
            onclick: node.attrs.onclick
        };
        node.attrs.route = undefined;
        if (node.attrs.href) { 
            log.info(`href ${node.attrs.href}`);
            attrs.href = node.attrs.href;
            attrs.target = attrs.target || '_blank';
            attrs.oncreate = m.route.link;
            attrs.onupdate = m.route.link;
            // attrs.onclick = () => window.open(node.attrs.href, '_blank');
            return m(`a.hs-layout ${css} ${this.getCSS(node)}`, attrs, content.map((c:any) => c));
        } else {
            return m(`.hs-layout ${css} ${this.getCSS(node)}`, attrs, content.map((c:any) => c));
        }
    }
}
