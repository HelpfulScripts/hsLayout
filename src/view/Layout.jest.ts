import { Layout } from "./Layout";
import { px } from "./Tokens";
import { m } from "../mithril";

let cont: any;
const root = window.document.createElement("div");

const layout = (type: string, contSizes: any[]) =>
  new Promise(resolve => {
    const config = {
      css: "myRow",
      content: [`First: ${contSizes[0]}`, "Second: remainder"],
      href: "my/link"
    };
    config[type] = contSizes;
    m.mount(root, { view: () => m(Layout, config) });
    cont = root;
    resolve();
  });

describe("Layout", () => {
  const contSizes = ["150px", "fill"];
  beforeAll(() => layout("rows", contSizes));

  it("should exist", () => {
    expect(Layout).toBeDefined();
  });

  it("should be clickable", () => {
    return expect(cont).toMatchInlineSnapshot(`
<div>
  <div
    class="hs-layout  myRow"
    href="#!my/link"
  >
    <div
      class="hs-layout  "
    >
      <div
        class="hs-leaf"
      >
        First: 150px
      </div>
    </div>
    <div
      class="hs-layout  "
    >
      <div
        class="hs-leaf"
      >
        Second: remainder
      </div>
    </div>
  </div>
</div>
`);
  });
});

describe("Layout without params", () => {
  const contSizes = [px(150), "fill"];
  beforeAll(() => layout("rows", contSizes));

  it("should exist", () => {
    expect(Layout).toBeDefined();
  });

  it("should be clickable", () => {
    return expect(cont).toMatchInlineSnapshot(`
<div>
  <div
    class="hs-layout  myRow"
    href="#!my/link"
  >
    <div
      class="hs-layout  "
    >
      <div
        class="hs-leaf"
      >
        First: [object Object]
      </div>
    </div>
    <div
      class="hs-layout  "
    >
      <div
        class="hs-leaf"
      >
        Second: remainder
      </div>
    </div>
  </div>
</div>
`);
  });
});
