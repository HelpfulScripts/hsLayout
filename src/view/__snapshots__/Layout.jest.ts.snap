// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should be clickable 1`] = `
<div>
  <a
    class="hs-layout  hs-row-layout myRow"
    href="#!my/link"
    target="_blank"
  >
    <div
      class="hs-layout  "
      style="width: 100%; top: 0px; height: 150px;"
    >
      <div
        class="hs-leaf"
      >
        First: 150px
      </div>
    </div>
    <div
      class="hs-layout  "
      style="width: 100%; top: 150px; bottom: 0px; height: auto;"
    >
      <div
        class="hs-leaf"
      >
        Second: remainder
      </div>
    </div>
  </a>
</div>
`;

exports[`Layout without params should be clickable 1`] = `
<div>
  <a
    class="hs-layout  hs-row-layout myRow"
    href="#!my/link"
    target="_blank"
  >
    <div
      class="hs-layout  "
      style="width: 100%; top: 0px; height: 150px;"
    >
      <div
        class="hs-leaf"
      >
        First: [object Object]
      </div>
    </div>
    <div
      class="hs-layout  "
      style="width: 100%; top: 150px; bottom: 0px; height: auto;"
    >
      <div
        class="hs-leaf"
      >
        Second: remainder
      </div>
    </div>
  </a>
</div>
`;
